library/src/blas1/rocblas_nrm2.hpp

```diff
@@ -0,0 +1,58 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#pragma once
+
+#include "fetch_template.h"
+#include "handle.h"
+#include "reduction_strided_batched.h"
+#include "rocblas.h"
+
+template <class To>
+struct rocblas_fetch_nrm2
+{
+    template <class Ti>
+    __forceinline__ __device__ To operator()(Ti x, ptrdiff_t tid)
+    {
+        return {fetch_abs2(x)};
+    }
+};
+
+struct rocblas_finalize_nrm2
+{
+    template <class To>
+    __forceinline__ __host__ __device__ To operator()(To x)
+    {
+        return sqrt(x);
+    }
+};
+
+// allocate workspace inside this API
+template <rocblas_int NB, typename Ti, typename To>
+rocblas_status rocblas_nrm2_template(rocblas_handle handle,
+                                     rocblas_int    n,
+                                     const Ti*      x,
+                                     rocblas_int    shiftx,
+                                     rocblas_int    incx,
+                                     To*            workspace,
+                                     To*            result)
+{
```

Comment:
The same comment here as for asum. I don't think there's a problem in combining the templates, but if you have a concern let me know!

Reply:
same response as asum.   There is no problem to combine.

Reply:
There are many ways to combine them. I am speaking generally here, but you could add an extra template parameter which passes a class, or an extra template template parameter which passes a template of classes. If you're just trying to distinguish between `asum` and `nrm2`, you only need one type parameter to differentiate them -- if more than one type or function needs to be passed for the different cases, then they can be wrapped in a class, and only the class is passed. I would prefer this method to using an `enum` to differentiate the functions, because an `enum` requires `constexpr if` or other tricks, and we don't use C++17 yet.

The fetch, reduce, and finalize functor classes can be enclosed in a class, and then that class passed to choose `asum`, `nrm2`, etc.

Reply:
I don't think Daine or I was suggesting merging the asum, and nrm2.   I don't think it saves much and keeping the functions separate is nice abstraction.    I think one could have merged a lot the framework into a C++ class and avoided all the duplicate patterns, but that might be another discussion.